### DISCUSSION TITLE
Cascade delate when using backlink relationship

### DISCUSSION THREAD

**Discussion by vardansargsyan92** [realm/realm-dotnet#2553] [Q&A] [answered]
Hello.

I have a backlink relationship through two objects inside my app

```
public class JobwalkItem : RealmObject
    {
        
        [PrimaryKey]
        [MapTo("_id")]
        public string Id { get; set; } = Guid.NewGuid().ToString();

        [Required]
        [MapTo("_partitionKey")]
        public string PartitionKey { get; set; }
  
        public int Count { get; set; }

        public double UsePeriod { get; set; }

        public string Name { get; set; }

        [Backlink(nameof(JobSummaryItem.JobwalkItem))]
        public IQueryable<JobSummaryItem> JobSummaryItems { get; }

    }
```
```
public class JobSummaryItem : RealmObject
    {
        [PrimaryKey]
        [MapTo("_id")]
        public string Id { get; set; } = Guid.NewGuid().ToString();

        [Required]
        [MapTo("_partitionKey")]
        public string PartitionKey { get; set; }

        public JobwalkItem JobwalkItem { get; set; }

        public JobSummaryItem ParentItem { get; set; }

        public bool HasChildren { get; set; }

        
        public int ItemCount { get; set; }

        public double TotalCubes { get; set; }

        public int TotalMoveLaborTime { get; set; }

        
        public double TotalUsePeriod { get; set; }

    }
```
When I delete a JobwalkItem I assume that it’s JobSummary items also should delete from database. But this does not happen and I need to manually delete all related JobSummary items and then delete JobwalkItem.

So the question is following. Is that an expected behavior and the cascade deletion is applicable only for embedded objects?

p.s I guess that cascade deletion is not unequivocally when we have backlinks.

**Comment by DominicFrei** [comment, marked as answer]
Hi @vardansargsyan92,

regarding cascading deletes there is no difference between forward links and backward links.
They don't automatically do cascading deletes unless you use `EmbeddedObject` as you already expected.

The reason behind this is that any `JobSummaryItem` in the `IQueryable<JobSummaryItem> JobSummaryItems` in your `JobwalkItem` could also be linked in any other `RealmObject`. We therefore cannot delete them safely.
Even if they weren't used anywhere else they are designed to not be deleted since they could be used somewhere else in the future.

The idea of having an object belonging to exactly just one other object and if it (or rather: the link to it) gets deleted there the object can be removed from the database is exactly what you would express by `EmbeddedObject`.
Cascading deletes are one of the main reasons `EmbeddedObject` was added in the first place, because it offers exactly that feature.

Does that answer your questions? Feel free to ask for clarification if you need any. 👍

**Reply by vardansargsyan92** [reply]
Thanks a lot for answer. Actually that was that I anticipated to see. It's a little hard to explain but in my code I have a case when the backlink collection deletes from database when I delete the "Parent" object. I know that is strange behavior  and that is why I asked the question. I guess I did something wrong there.

**Reply by DominicFrei** [reply]
Glad I could help.  Since I don't know your code I can only guess but: would it help to turn that relationship around and use an `EmbeddedObject` instead? Or is it easier for you to implement the delete yourself?

**Reply by vardansargsyan92** [reply]
Actually that object is a little more than Embedded Object, so should be keep as it was. The joke was that I had a functionality implementing cascade deletion. During previous PR we decided to remove that code because we noticed that objects were delete when delete the "Parent" object. Then I faced up with the same story but now the backlink objects would not be deleted when delete corresponding object. So that drive me crazy))). So I decided that it would be better to ask you guys what is correct what is wrong

**Reply by DominicFrei** [reply]
Ah, that sounds like an interesting back and forth. :) Oftentimes we find the perfect solution going step by step and re-designing things. I can totally understand how you ended up there. Glad you found a solution in the end that suited your needs. 👍
